applications/FluidDynamicsApplication/custom_processes/mass_conservation_check_process.cpp

```diff
@@ -0,0 +1,265 @@
+                std::ofstream myfile;
+                myfile.open ("volumeLog.txt", std::ios::out | std::ios::app);
+                if ( myfile.is_open() ){
+
+                    myfile << mCounter << "    ";
+                    myfile << posVol << "    ";
+                    if ( mCompareToInitial ){ 
+                        myfile << posVol / mInitialPositiveVolume * 100.0 << "    "; 
+                    }
+                    myfile << negVol << "    ";
+                    if ( mCompareToInitial ){ 
+                        myfile << negVol / mInitialNegativeVolume * 100.0 << "\n"; 
+                    } 
+                    myfile.close();
+                }
+            }
+        }
+
+        mCounter++;
+    } else {
+        return;
+    }
+};
+
+
+/* Private functions ****************************************************/
+
+void MassConservationCheckProcess::ComputeVolumesOfFluids( double &positiveVolume, double &negativeVolume )
+{
+    // useless containers
+    Matrix rShapeFunctionsPos, rShapeFunctionsNeg;
+    GeometryType::ShapeFunctionsGradientsType rShapeDerivativesPos, rShapeDerivativesNeg;
+    Vector null3(3, 0.0);
+    Vector null4(4, 0.0);
+
+    // abbreviation of variable names
+    double Vol;
+    double& posVol = positiveVolume;
+    double& negVol = negativeVolume;
+
+    // initalisation
+    posVol = 0.0;
+    negVol = 0.0;
+
+    for (int i_elem = 0; i_elem < static_cast<int>(mrModelPart.NumberOfElements()); ++i_elem){
+        // iteration over all elements
+        auto it_elem = mrModelPart.ElementsBegin() + i_elem;
+        
+        auto p_geom = it_elem->pGetGeometry();
+        unsigned int ptCountPos = 0;
+        unsigned int ptCountNeg = 0;
+
+        // instead of using data.isCut()
+        for (unsigned int pt = 0; pt < p_geom->Points().size(); pt++){
+            if ( p_geom->GetPoint(pt).FastGetSolutionStepValue(DISTANCE) > 0.0 ){
+                ptCountPos++;
+            } else {
+                ptCountNeg++;
+            }
+        }
+
+        if ( ptCountPos == p_geom->Points().size() ){
+            // all nodes are positive
+            if( p_geom->Points().size() == 3 ){
+                Vol = p_geom->Area();
+            } else {
+                Vol = p_geom->Volume();
+            }
+            posVol += Vol;
+        } 
+        else if ( ptCountNeg == p_geom->Points().size() ){
+            // all nodes are negative
+            if( p_geom->Points().size() == 3 ){
+                Vol = p_geom->Area();
+            } else {
+                Vol = p_geom->Volume();
+            }
+            negVol += Vol;
+        }
+        else {
+            // element is cut by the surface
+            ModifiedShapeFunctions::Pointer p_modified_sh_func = nullptr;
+            Vector w_gauss_pos_side(3, 0.0);
+            Vector w_gauss_neg_side(3, 0.0);
+
+            // the element is cut by the interfae between the fluids
+            if( p_geom->Points().size() == 3 ){
```

Comment:
You can avoid this if statement by resizing distance according to geometries `PointsNumber()` and by doing the loop from zero to such value. Something like
```
for (unsigned int i_node = 0; i_node < p_geom->PointsNumber(); ++i_node) {
    ..........
}
```

Reply:
Done.